### PR TITLE
docstring for @inferred

### PR DESCRIPTION
**Pull request by felipenoris**

### CODE REVIEW

**Comment by felipenoris** [comment]
CI doesn't like it. But I guess there's nothing wrong.

**Comment by StefanKarpinski** [comment]
I cancelled the CI since it's just a doc change.

**Comment by tkelman** [comment]
Still need to verify whitespace to avoid merging something to master that will then cause every other PR to fail.

**Comment by felipenoris** [comment]
@tkelman this was not the case with this one, right?

**Comment by tkelman** [comment]
No don't think so, but it's a risk until we resolve #11830 and just have a bot fix whitespace ~weekly

**Comment by felipenoris** [comment]
You´re right.

**Comment by andyferris** [comment]
Is this docstring correct in saying that we can't use `@inferred` on calls *to* generated functions? In reflection.jl, it seems we can't make a call to `return_types` *from* a generated function (@simonster seems to be the author of `@inferred` so I'll ping you too).

I'm only asking because I just discovered this feature (and I love it!) and have been using it on lots of generated functions and was wondering if my tests were authentic or not.

**Comment by felipenoris** [comment]
```julia
function return_types(f::ANY, types::ANY=Tuple)
    ccall(:jl_is_in_pure_context, Bool, ()) && error("code reflection cannot be used from generated functions")
```

**Comment by felipenoris** [comment]
So the correct doc should be:
```
`f(x)` can be any call expression. This macro cannot be called from a generated functions.
```
Right?

**Comment by andyferris** [comment]
Yes, that's my interpretation. Thanks for the docstring, btw. I only discovered this a few days ago and had to experiment a bit to get it doing what I wanted.

If fact, sometimes I wanted to check if a function was inferred without actually performing a test on values of the output, so I defined an additional macro:

```julia
# @inferred throws an error, which doesn't interract particularly well
# with BaseTestNext
macro test_inferred(ex)
    quote
        @test try
            @inferred($ex)
            true
        catch
            false
        end
    end
end
```

This lets me run a test for the correct value and a separate test for inference (or just inference).

**Comment by andyferris** [comment]
Also, we should probably also indicate if that means inside a generated function's code generation stage or output code stage (I suspect the former, but I'm not sure what `jl_is_in_pure_context` exactly is...)

**Comment by vtjnash** [comment]
I think it's better to list those restrictions under "writing a `@generated` function" since the generator part is not allowed to do very much (it must be pure).

**Comment by felipenoris** [comment]
For now I just removed the reference to generated functions. Later I can suggest some text to the "writing a `@generated` function" doc. https://github.com/JuliaLang/julia/pull/16497

File: base/test.jl [hunk 1]
```diff
@@ -782,6 +782,49 @@ macro test_approx_eq(a, b)
...
+
+julia> @code_warntype f(1,2,3)
+Variables:
+  #self#::#f
+  a::Int64
+  b::Int64
+  c::Int64
+
+Body:
+  begin  # REPL[2], line 1:
```
Comment (tkelman) [inline base/test.jl]: did you verify that this newly added docstring actually passes? I think you may need some `<BLANKLINE>` markers and to fix a few of the line numbers etc to get sphinx to pass

Comment (felipenoris) [inline base/test.jl]: I didn´t. I had a few problems with that. I´ll work on that now.

Comment (tkelman) [inline base/test.jl]: see 632e0873574a04b3b3e734f53feec53814826d31, not sure how much of that is accurate for master too

Comment (felipenoris) [inline base/test.jl]: @tkelman Thanks! I´ll check it out.
I had problems to run `make doctest` on my mac. I´m setting up a Linux VM to run it.

Comment (tkelman) [inline base/test.jl]: there are a lot of failures on master, I don't expect you to fix them all but at least get this one you added to pass

**Pull request merged**